File: src/main/java/com/github/cao/awa/kalmia/message/database/message/MessageDatabase.java
package com.github.cao.awa.kalmia.message.database.message;

import com.github.cao.awa.apricot.io.bytes.reader.BytesReader;
import com.github.cao.awa.apricot.util.collection.ApricotCollectionFactor;
import com.github.cao.awa.kalmia.annotations.number.encode.ShouldSkipped;
import com.github.cao.awa.kalmia.database.KeyValueBytesDatabase;
import com.github.cao.awa.kalmia.database.provider.DatabaseProviders;
import com.github.cao.awa.kalmia.mathematic.base.SkippedBase256;
import com.github.cao.awa.kalmia.message.DeletedMessage;
import com.github.cao.awa.kalmia.message.Message;
import com.github.cao.awa.kalmia.message.unknown.UnknownMessage;
import com.github.cao.awa.viburnum.util.bytes.BytesUtil;
import com.github.zhuaidadaya.rikaishinikui.handler.universal.entrust.EntrustEnvironment;

import java.util.Set;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

public class MessageDatabase {
    private final KeyValueBytesDatabase database;

    public MessageDatabase(String path) throws Exception {
        this.database = DatabaseProviders.bytes(path);
    }

    public void operation(@ShouldSkipped byte[] sid, BiConsumer<Long, Message> action) {
        seqAll(sid,
               seq -> {
                   action.accept(seq,
                                 get(sid,
                                     SkippedBase256.longToBuf(seq)
                                 )
                   );
               }
        );
    }

    public void operation(@ShouldSkipped byte[] sid, long from, long to, BiConsumer<Long, Message> action) {
        seqAll(sid,
               from,
               to,
               seq -> {
                   action.accept(seq,
                                 get(sid,
                                     SkippedBase256.longToBuf(seq)
                                 )
                   );
               }
        );
    }

    public Message get(@ShouldSkipped byte[] sid, @ShouldSkipped byte[] seq) {
        byte[] msgBytes = this.database.get(key(sid,
                                                seq
        ));
        if (msgBytes == null) {
            return null;
        }
        return EntrustEnvironment.trys(() -> Message.create(msgBytes),
                                       () -> new UnknownMessage(msgBytes)
        );
    }

    public void delete(@ShouldSkipped byte[] sid, @ShouldSkipped byte[] seq) {
        Message source = get(sid,
                             seq
        );
        this.database.put(key(sid,
                              seq
                          ),
                          new DeletedMessage(source.getSender(),
                                             source.digest()
                          ).toBytes()
        );
    }

    public void seqAll(@ShouldSkipped byte[] sid, Consumer<Long> action) {
        byte[] seqByte = this.database.get(sid);

        long count = seqByte == null ? - 1 : SkippedBase256.readLong(BytesReader.of(seqByte));

        count++;

        if (count > 0) {
            for (long seq = 0; ; seq++) {
                if (seq > count) {
                    break;
                }
                action.accept(seq);
            }
        }
    }

    public void seqAll(@ShouldSkipped byte[] sid, long from, long to, Consumer<Long> action) {
        long count = curSeq(sid);

        if (count > 0) {
            long endIndex = Math.min(to,
                                     count
            );
            for (long seq = from; ; seq++) {
                if (seq > endIndex) {
                    break;
                }
                action.accept(seq);
            }
        }
    }

    public boolean present(byte[] key) {
        return this.database.get(key) != null;
    }

    public long seq(@ShouldSkipped byte[] sid) {
        byte[] seqByte = this.database.get(sid);

        return seqByte == null ? - 1 : SkippedBase256.readLong(BytesReader.of(seqByte));
    }

    public long curSeq(@ShouldSkipped byte[] sid) {
        byte[] seqByte = this.database.get(sid);

        return seqByte == null ? 0 : SkippedBase256.readLong(BytesReader.of(seqByte)) + 1;
    }

    public void deleteAll(@ShouldSkipped byte[] sid) {
        seqAll(sid,
               seq -> {
                   delete(sid,
                          SkippedBase256.longToBuf(seq)
                   );
               }
        );
    }

    public byte[] key(@ShouldSkipped byte[] sid, @ShouldSkipped byte[] seq) {
        return BytesUtil.concat(sid,
                                seq
        );
    }

    public Set<Long> search(@ShouldSkipped byte[] sid, String target) {
        Set<Long> result = ApricotCollectionFactor.hashSet();

        operation(sid,
                  (seq, msg) -> {
//            if (msg.contains(target)) {
//                result.add(seq);
//            }
                  }
        );

        return result;
    }

    public long send(@ShouldSkipped byte[] sid, Message msg) {
        byte[] seqByte = this.database.get(sid);

        long seq = seqByte == null ? - 1 : SkippedBase256.readLong(BytesReader.of(seqByte));

        long nextSeq = seq + 1;

        byte[] nextSeqByte = SkippedBase256.longToBuf(nextSeq);

        this.database.put(key(sid,
                              nextSeqByte
                          ),
                          msg.toBytes()
        );

        this.database.put(sid,
                          nextSeqByte
        );

        return nextSeq;
    }
}

File: src/main/java/com/github/cao/awa/kalmia/message/database/message/MessageDatabase.kt
package com.github.cao.awa.kalmia.message.database.message

import com.github.cao.awa.apricot.io.bytes.reader.BytesReader
import com.github.cao.awa.apricot.util.collection.ApricotCollectionFactor
import com.github.cao.awa.kalmia.annotations.number.encode.ShouldSkipped
import com.github.cao.awa.kalmia.database.KeyValueBytesDatabase
import com.github.cao.awa.kalmia.database.KeyValueDatabase
import com.github.cao.awa.kalmia.database.key.BytesKey
import com.github.cao.awa.kalmia.database.provider.DatabaseProviders
import com.github.cao.awa.kalmia.identity.LongAndExtraIdentity
import com.github.cao.awa.kalmia.identity.PureExtraIdentity
import com.github.cao.awa.kalmia.mathematic.base.SkippedBase256
import com.github.cao.awa.kalmia.message.Message
import com.github.cao.awa.kalmia.message.deleted.DeletedMessage
import com.github.cao.awa.viburnum.util.bytes.BytesUtil
import java.util.function.BiConsumer
import java.util.function.Consumer

class MessageDatabase(path: String) : KeyValueDatabase<BytesKey, Message?>(ApricotCollectionFactor::hashMap) {
    private val delegate: KeyValueBytesDatabase = DatabaseProviders.bytes(path)

    fun operation(@ShouldSkipped sessionIdentity: PureExtraIdentity, action: BiConsumer<Long, Message?>) {
        seqAll(
            sessionIdentity
        ) {
            action.accept(
                it,
                get(
                    sessionIdentity,
                    BytesKey(SkippedBase256.longToBuf(it))
                )
            )
        }
    }

    fun operation(
        @ShouldSkipped sessionIdentity: PureExtraIdentity,
        from: Long,
        to: Long,
        action: BiConsumer<Long, Message?>
    ) {
        seqAll(
            sessionIdentity,
            from,
            to
        ) {
            action.accept(
                it,
                get(
                    sessionIdentity,
                    BytesKey(SkippedBase256.longToBuf(it))
                )
            )
        }
    }

    override operator fun get(key1: BytesKey, @ShouldSkipped key2: BytesKey): Message? {
        return this[
            identity(
                key(
                    key1,
                    key2
                )
            )
        ]
    }

    operator fun get(sessionIdentity: PureExtraIdentity, @ShouldSkipped seq: BytesKey): Message? {
        return this[BytesKey(sessionIdentity.extras()), seq]
    }

    override operator fun get(key: BytesKey): Message? {
        return cache()[
            key,
            { getMessage(it) }
        ]
    }

    operator fun get(identity: LongAndExtraIdentity): Message? {
        return this[BytesKey(identity.toBytes())]
    }

    private fun getMessage(identity: BytesKey): Message? {
        val data = this.delegate[identity] ?: return null
        return Message.create(data)
    }

    fun getMessage(identity: LongAndExtraIdentity): Message? {
        return getMessage(BytesKey(identity.toBytes()))
    }

    override fun remove(key: BytesKey) {
        cache().delete(
            key
        ) {
            this.delegate.remove(it)
        }
    }

    fun remove(identity: LongAndExtraIdentity) {
        remove(BytesKey(identity.toBytes()))
    }

    override fun remove(@ShouldSkipped key1: BytesKey, @ShouldSkipped key2: BytesKey) {
        val key = key(
            key1,
            key2
        )
        remove(identity(key))
        this.delegate.remove(key)
    }

    fun markDelete(identity: LongAndExtraIdentity) {
        val source = get(identity) ?: return
        this[BytesKey(identity.toBytes())] = DeletedMessage(
            source.sender(),
            source.digest()
        )
    }

    fun markDelete(sessionIdentity: PureExtraIdentity, @ShouldSkipped seq: BytesKey) {
        markDelete(
            identity(
                key(
                    sessionIdentity,
                    seq
                )
            )
        )
    }

    fun identity(key: BytesKey): LongAndExtraIdentity {
        return LongAndExtraIdentity.read(BytesReader.of(this.delegate[key]))
    }

    fun identity(sessionIdentity: PureExtraIdentity, seq: BytesKey, messageIdentity: LongAndExtraIdentity) {
        this.delegate[key(sessionIdentity, seq)] = messageIdentity.toBytes()
    }

    fun seqAll(@ShouldSkipped sessionIdentity: PureExtraIdentity, action: Consumer<Long>) {
        val count = nextSeq(sessionIdentity) - 1
        if (count > 0) {
            var seq: Long = 0
            while (true) {
                if (seq > count) {
                    break
                }
                action.accept(seq)
                seq++
            }
        }
    }

    fun seqAll(sessionIdentity: PureExtraIdentity, from: Long, to: Long, action: Consumer<Long>) {
        val count = nextSeq(sessionIdentity) - 1
        if (count > 0) {
            val endIndex = Math.min(
                to,
                count
            )
            var seq = from
            while (true) {
                if (seq > endIndex) {
                    break
                }
                action.accept(seq)
                seq++
            }
        }
    }

    fun nextSeq(sessionIdentity: PureExtraIdentity): Long = curSeq(sessionIdentity) + 1

    fun curSeq(sessionIdentity: PureExtraIdentity): Long {
        val seqByte = this.delegate[BytesKey(sessionIdentity.extras())]
        return if (seqByte == null) -1 else SkippedBase256.readLong(BytesReader.of(seqByte))
    }

    fun curSeq(sessionIdentity: PureExtraIdentity, @ShouldSkipped seq: ByteArray) {
        this.delegate[BytesKey(sessionIdentity.extras())] = seq
    }

    fun deleteAll(sessionIdentity: PureExtraIdentity) {
        seqAll(
            sessionIdentity
        ) {
            markDelete(
                sessionIdentity,
                BytesKey(SkippedBase256.longToBuf(it))
            )
        }
    }

    fun key(sessionIdentity: BytesKey, @ShouldSkipped seq: BytesKey): BytesKey {
        return BytesKey(
            BytesUtil.concat(
                sessionIdentity.key(),
                seq.key()
            )
        )
    }

    fun key(sessionIdentity: PureExtraIdentity, @ShouldSkipped seq: BytesKey): BytesKey {
        return key(BytesKey(sessionIdentity.extras()), seq)
    }

    fun search(sessionIdentity: PureExtraIdentity): Set<Long> {
        val result: Set<Long> = ApricotCollectionFactor.hashSet()
        operation(
            sessionIdentity
        ) { _, _ -> }
        return result
    }

    override fun set(key: BytesKey, value: Message?) {
        cache().update(
            key,
            value,
            this::update
        )
    }

    fun set(identity: LongAndExtraIdentity, msg: Message) {
        this[BytesKey(identity.toBytes())] = msg
    }

    private fun update(identity: BytesKey, msg: Message?) {
        if (msg == null) {
            this.remove(identity)
        } else {
            this.delegate[identity] = msg.toBytes()
        }
    }

    fun send(sessionIdentity: PureExtraIdentity, msg: Message): Long {
        val nextSeq = nextSeq(sessionIdentity)
        val nextSeqByte = SkippedBase256.longToBuf(nextSeq)

        // Save the redirect to global id.
        identity(
            sessionIdentity,
            BytesKey(nextSeqByte),
            msg.identity()
        )

        // Update index.
        curSeq(
            sessionIdentity,
            nextSeqByte
        )

        // Update message.
        set(
            msg.identity(),
            msg
        )

        // Return two identity result.
        return nextSeq
    }
}